File: rust/src/code/number_of_islands.rs
pub struct Solution {}

use std::collections::HashSet;
impl Solution {
    // NOTE: time complexity: O(n * m)
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let mut count: i32 = 0;
        let mut visited: HashSet<(usize, usize)> = HashSet::new();
        for i in 0..grid.len() {
            for j in 0..grid[i].len() {
                if grid[i][j] != '0' && !visited.contains(&(i, j)) {
                    count += 1;
                    Self::dfs(i, j, &mut visited, &grid);
                }
            }
        }
        count
    }

    fn dfs(i: usize, j: usize, visited: &mut HashSet<(usize, usize)>, grid: &Vec<Vec<char>>) {
        if i >= grid.len() || j >= grid[i].len() || grid[i][j] == '0' || visited.contains(&(i, j)) {
            return;
        }
        visited.insert((i, j));
        Self::dfs(i + 1, j, visited, grid);
        if i.checked_sub(1).is_some() {
            Self::dfs(i - 1, j, visited, grid);
        }
        Self::dfs(i, j + 1, visited, grid);
        if j.checked_sub(1).is_some() {
            Self::dfs(i, j - 1, visited, grid);
        }
    }
}

#[cfg(test)]
pub mod test {
    use super::Solution;

    #[test]
    fn case_1() {
        // arrange
        let grid: Vec<Vec<char>> = vec![
            vec!['1', '1', '1', '1', '0'],
            vec!['1', '1', '0', '1', '0'],
            vec!['1', '1', '0', '0', '0'],
            vec!['0', '0', '0', '0', '0'],
        ];
        let expected = 1;

        // act
        let actual = Solution::num_islands(grid);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn case_2() {
        // arrange
        let grid: Vec<Vec<char>> = vec![
            vec!['1', '1', '0', '0', '0'],
            vec!['1', '1', '0', '0', '0'],
            vec!['0', '0', '1', '0', '0'],
            vec!['0', '0', '0', '1', '1'],
        ];
        let expected = 3;

        // act
        let actual = Solution::num_islands(grid);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn t_case_3() {
        // arrange
        let grid: Vec<Vec<char>> = vec![vec!['1'], vec!['1']];
        let expected = 1;

        // act
        let actual = Solution::num_islands(grid);

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/number_of_islands.py
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        """time complexity: O(n * m), space complexity: O(n * m)"""
        result: int = 0
        visited: set[tuple[int, int]] = set()
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if (i, j) in visited or grid[i][j] == "0":
                    continue
                self.dfs(i, j, visited, grid)
                result += 1
        return result

    def dfs(
        self, row: int, col: int, visited: set[tuple[int, int]], grid: [list[list[str]]]
    ) -> None:
        if (
            row < 0
            or row >= len(grid)
            or col < 0
            or col >= len(grid[row])
            or (row, col) in visited
            or grid[row][col] == "0"
        ):
            return
        visited.add((row, col))
        self.dfs(row + 1, col, visited, grid)
        self.dfs(row - 1, col, visited, grid)
        self.dfs(row, col + 1, visited, grid)
        self.dfs(row, col - 1, visited, grid)


def test_numIslands_case_1():
    # arrange
    grid: list[list[str]] = [
        ["1", "1", "1", "1", "0"],
        ["1", "1", "0", "1", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "0", "0", "0"],
    ]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.numIslands(grid)

    # assert
    assert actual == expected


def test_numIslands_case_2():
    # arrange
    grid: list[list[str]] = [
        ["1", "1", "0", "0", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "1", "0", "0"],
        ["0", "0", "0", "1", "1"],
    ]
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.numIslands(grid)

    # assert
    assert actual == expected
